servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/IdleTimeoutConnectionFilter.java

```diff
@@ -0,0 +1,242 @@
+        @Nullable
+        private volatile Cancellable timeoutTask;
+
+        private final long timeoutNs;
+        private final Executor timeoutExecutor;
+
+        // While it may look like "volatile" is not required for this variable for "happens-before" visibility because
+        // we always write to "requests" after updating "lastResponseTime" and read "requests" before reading
+        // "lastResponseTime", non-volatile writes to "long" variables are not atomic and may result in incorrect
+        // reading of the value when 2 threads race.
+        private volatile long lastResponseTime;
+
+        ConnectionIdleTimeoutFilterImpl(final FilterableStreamingHttpConnection connection,
+                                        final long timeoutNs,
+                                        final Executor timeoutExecutor) {
+            super(connection);
+            this.timeoutNs = timeoutNs;
+            this.timeoutExecutor = timeoutExecutor;
+            connection.onClose().whenFinally(this::cancelTask).subscribe();
+            this.lastResponseTime = nanoTime();
+            timeoutTask = this.timeoutExecutor.schedule(this, timeoutNs, NANOSECONDS);
+        }
+
+        private long nanoTime() {
+            return timeoutExecutor.currentTime(NANOSECONDS);
+        }
+
+        private void cancelTask() {
+            final Cancellable oldTask = timeoutTaskUpdater.getAndSet(this, CANCELLED);
+            if (oldTask != null) {
+                oldTask.cancel();
+            }
+        }
+
+        @Override
+        public Single<StreamingHttpResponse> request(final StreamingHttpRequest request) {
+            return defer(() -> {
+                final int inFlightRequests = requestsUpdater.incrementAndGet(this);
+                if (inFlightRequests < 0) {
+                    // To mitigate the tiny risk of going into positive numbers with Integer.MAX_VALUE concurrent
+                    // requests (unrealistic), reset the counter back to MIN_VALUE.
+                    requests = MIN_VALUE;
+                    return failed(new RetryableClosedChannelException(delegate(), timeoutNs));
+                }
+                return delegate().request(request)
+                        .liftSync(new BeforeFinallyHttpOperator(() -> {
+                            // It's acceptable to use 2 volatile variables instead of a single object state here. Even
+                            // if 2 threads race between updating "lastResponseTime" and "requests", the delay for a new
+                            // timer task will be close to "timeoutNs".
+                            requestsUpdater.updateAndGet(this, prev -> {
+                                if (prev > 1) {
+                                    return prev - 1;
+                                }
+                                assert prev == 1 : "Unexpected requests value: " + prev;
+                                lastResponseTime = nanoTime();
```

Comment:
the updater function maybe called repeated times if the atomic operation fails, and doesn't ensure the count is `0` ... this should be moved outside the scope of the updater.

Reply:
Assuming there is no overflow: the increment and decrement are tight to a request. It can go to 0 only if all requests complete. It may race between N to N-1 transition but it will never race from 1 to 0.

Reply:
This code looks better 👍 .The issue was updating `lastResponseTime` state was inside the updater function (which actually writes before `requestsUpdater` is updated and maybe invoked multiple times if the atomic operation fails).